### PR TITLE
22 item availability

### PR DESCRIPTION
**Pull request by FV89**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -17,14 +17,16 @@ class Item(BaseModel):
...
 
     def json(self):
         return {
             'uuid': str(self.uuid),
             'name': self.name,
             'price': int(self.price),
             'description': self.description,
-            'category': self.category
+            'category': self.category,
+            'availability': self.availability
```
Comment (GendoIkari) [inline models.py]: SI cerca sempre di terminare gli elenchi con una `,` anche se e' l'ultimo elemento.

```'availability': self.availability,```

File: tests/test_item.py [hunk 2]
```diff
@@ -30,16 +30,18 @@ def test_get_items(self):
             uuid=uuid.uuid4(),
             name='Item one',
             price=5,
-            description='Description one',
-            category='Category one'
+            description='dhfsdjòfgjasdògj',
+            category='poligoni',
+            availability=8
```
Comment (GendoIkari) [inline tests/test_item.py]: Errore di rebase probabilmente, ripristiniamo i nomi sensati invece di quelli incomprensibili.

File: tests/test_item.py [hunk 3]
```diff
@@ -30,16 +30,18 @@ def test_get_items(self):
...
 
         item2 = Item.create(
             uuid=uuid.uuid4(),
             name='Item two',
             price=15,
-            description='Description two',
-            category='Category two'
+            description='desc2',
+            category='smartphone',
+            availability=11
```
Comment (GendoIkari) [inline tests/test_item.py]: Questi numeri casuali non vanno molto bene, cerchiamo di evitare la casualita' fuorviante nel codice. DIamo un numero sensato uguale per tutti (tranne quando stai testando i corner case...)

File: tests/test_item.py [hunk 4]
```diff
@@ -80,9 +83,10 @@ def test_create_item__failure_empty_field(self):
     def test_create_item__failure_empty_field_only_spaces(self):
         new_item_data = {
             'name': '    ',
-            'price': 10,
-            'description': 'Description one',
-            'category': 'Category one'
+            'price': 123,
```
Comment (GendoIkari) [inline tests/test_item.py]: Idem come prima, price 10 non andava bene? Perche;? Probabilmente si', errore di rebase forse.

File: tests/test_order.py [hunk 5]
```diff
@@ -27,16 +27,18 @@ def setup_class(cls):
...
         cls.item2 = Item.create(
             uuid=str(uuid.uuid4()),
             name='Item two',
             price=10,
             description='Item two description',
             category='Category one',
+            availability=17
         )
 
         app.config['TESTING'] = True
```
Comment (GendoIkari) [inline tests/test_order.py]: Non vedo test nuovi, quindi immagino che manchino i test per:
- codice di errore nella richiesta per creare ordini con availability negativa
- se provo a creare un item con availability negativa mi dia codice di errore

File: views/item.py [hunk 6]
```diff
@@ -19,6 +19,7 @@ def post(self):
         parser.add_argument('price', type=int, required=True)
         parser.add_argument('description', type=str, required=True)
         parser.add_argument('category', type=str, required=True)
+        parser.add_argument('availability', type=int, required=True)
```
Comment (GendoIkari) [inline views/item.py:22]: Qui non vedo check su availability. Potrebbe essere -9999 giusto? E questa sarebbe una BAD_REQUEST.

File: tests/test_order.py (hunk 5 above)
Comment (GendoIkari) [reply tests/test_order.py]: E non vedo assert nuovi, quindi in realta' non stai testando niente della tua feature.
Dovresti controllare l'availability corretta nei casi:
- creazione order
- modifica order
- cancellazione order

File: views/order.py [hunk 7]
```diff
@@ -91,8 +98,19 @@ def put(self, uuid):
...
+            if item_quantity > item.availability:
+                return None, BAD_REQUEST
+
         with database.transaction():
-            OrderItem.delete().where(OrderItem.order == order).execute()
+            temp_query = OrderItem.select().where(OrderItem.order == order.id)
+            for order_item in temp_query:
+                item_temp_query = Item.get(Item.id == order_item.item)
+                item_temp_query.availability = (item_temp_query.availability + order_item.quantity)
+                item_temp_query.save()
```
Comment (GendoIkari) [inline views/order.py]: Abbiamo le backref nei model (realated_names), usiamole.
`order_item.item` e' gia' l'Item che ti serve.

File: views/order.py [hunk 8]
```diff
@@ -115,8 +135,12 @@ def delete(self, uuid):
...
 
         with database.transaction():
-            order_items = OrderItem.select().where(
-                OrderItem.order == order)
+            order_items = OrderItem.select().where(OrderItem.order == order.id)
+
+            for order_item in order_items:
+                temp_query = Item.get(Item.id == order_item.item)
+                temp_query.availability = temp_query.availability + order_item.quantity
+                temp_query.save()
```
Comment (GendoIkari) [inline views/order.py]: Idem come prima. Le backref.

File: tests/test_order.py [hunk 9]
```diff
@@ -150,6 +170,24 @@ def test_modify_order__success(self):
         assert len(order1_items) == 1
         assert str(order1_items[0].item.uuid) == str(self.item2.uuid)
 
+        temp_item = Item.get(Item.uuid == self.item2.uuid)
+        assert temp_item.availability == (start_availability - 2)
```
Comment (GendoIkari) [inline tests/test_order.py:171]: `self.item2.availability`

File: tests/test_order.py [hunk 10]
```diff
@@ -127,6 +144,9 @@ def test_modify_order__success(self):
         order1 = self.create_order(self.user1)
         order2 = self.create_order(self.user1)
 
+        temp_item = Item.get(Item.uuid == self.item2.uuid)
+        start_availability = temp_item.availability
```
Comment (GendoIkari) [inline tests/test_order.py]: Di nuovo, queste get dovrebbero essere inutili.

File: tests/test_order.py [hunk 11]
```diff
@@ -80,6 +83,20 @@ def test_create_order__success(self):
         order_total = (self.item1.price * 2) + self.item2.price
         assert order_from_server['total_price'] == order_total
 
+        temp_item = Item.get(Item.uuid == self.item1.uuid)
```
Comment (GendoIkari) [inline tests/test_order.py]: Ancora.

File: tests/test_order.py [hunk 12]
```diff
@@ -53,6 +53,9 @@ def test_get_order__success(self):
         assert order_from_server == order.json()
 
     def test_create_order__success(self):
+        temp_item = Item.get(Item.uuid == self.item1.uuid)
```
Comment (GendoIkari) [inline tests/test_order.py]: Anche qui.
